File: src/tools/files.py
from pathlib import Path
from ..config import BASE_DIR


def file_read(path: str) -> str:
    """Read a small text file from the workspace."""
    p = (BASE_DIR / path).resolve()
    if not str(p).startswith(str(BASE_DIR)):
        raise ValueError("Path outside workspace")
    return p.read_text(encoding="utf-8")


def file_write(path: str, content: str) -> str:
    """Overwrite or create a text file in the workspace."""
    p = (BASE_DIR / path).resolve()
    if not str(p).startswith(str(BASE_DIR)):
        raise ValueError("Path outside workspace")
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(content, encoding="utf-8")
    return f"wrote {len(content)} chars to {path}"

File: common/src/tools/files.rs
/// File read/write tools — sandboxed to the workspace BASE_DIR.
use anyhow::{bail, Result};
use std::path::{Path, PathBuf};
use crate::config;

fn safe_path(rel: &str) -> Result<PathBuf> {
    let base = &config::get().base_dir;
    let resolved = base.join(rel).canonicalize()
        .unwrap_or_else(|_| base.join(rel));
    if !resolved.starts_with(base) {
        bail!("path outside workspace");
    }
    Ok(resolved)
}

pub fn file_read(path: &str) -> Result<String> {
    let p = safe_path(path)?;
    Ok(std::fs::read_to_string(p)?)
}

pub fn file_write(path: &str, content: &str) -> Result<String> {
    let p = safe_path(path)?;
    if let Some(parent) = p.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(&p, content)?;
    Ok(format!("wrote {} chars to {path}", content.len()))
}
